lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs

```diff
@@ -472,6 +491,66 @@ propDivvyFeeInvariantEmptyList (fee, outs) =
   where
     prop = divvyFee fee outs `seq` True
 
+{-------------------------------------------------------------------------------
+                         Fee Adjustment - properties
+-------------------------------------------------------------------------------}
+
+prop_rebalanceSelection
+    :: CoinSelection
+    -> OnDanglingChange
+    -> Property
+prop_rebalanceSelection sel onDangling = do
+    let (sel', fee') = rebalanceSelection opts sel
+    let selectionIsBalanced = case onDangling of
+            PayAndBalance ->
+                fee' /= Fee 0 || Fee (delta sel') == estimateFee opts sel'
+            SaveMoney ->
+                fee' /= Fee 0 || Fee (delta sel') >= estimateFee opts sel'
+    let reserveIsEmpty =
+            case reserve sel of
+                Nothing -> isNothing (reserve sel')
+                Just{}  -> reserve sel' == Just (Coin 0)
+    conjoin
+        [ property selectionIsBalanced
+        , property reserveIsEmpty
+        ]
+        & counterexample (unlines
+            [ "selection (before):", pretty sel
+            , "selection (after):", pretty sel'
+            , "delta (before): " <> show (delta sel)
+            , "delta (after):  " <> show (delta sel')
+            , "remaining fee:  " <> show (getFee fee')
+            ])
+        & classify (reserveNonNull && feeLargerThanDelta)
+            "reserve > 0 && fee > delta"
+        & classify (reserveLargerThanFee && feeLargerThanDelta)
+            "reserve > fee && fee > delta"
+        & classify reserveLargerThanFee
+            "reserve > fee"
+        & classify feeLargerThanDelta
+            "fee > delta"
```

Comment:
any chance to test `ErrValidateSelection` ? :thinking:

Reply:
In the fee spec, no chance, and in integration also quite hard. This is actually a server error (500), so something that we shouldn't be able to trigger under any circumstances, in theory.